### ISSUE TITLE
Make the Java monad and instance of Category

### ISSUE THREAD

**Issue by rahulmutt** [typelead/eta#687]
If the `Java` monad was an instance of Category, we can use `>>>` for composing fluent APIs instead of `>-`.

**Comment by aboelhamd** [comment]
Where could I start from , rahulmutt ?

**Comment by rahulmutt** [comment]
@aboelhamd The change will most likely be made in `libraries/base/Java/Core.hs` where you can specify the instance definition for `Category`. The implementations of the other functions in that module should be helpful in giving you an idea of how to work with the Java monad in its raw form.

**Comment by alanasp** [comment]
@rahulmutt Sorry for asking this but this is my first ETA issue and I just wanted some clarification on how everything's structured. I have been studying the codebase quite a bit but I don't seem to understand how the Java Monad is defined. I found out that it is an instance of Monad, Applicative and Functor but I have not found the definition in terms of type and value constructors, e.g.: "data Java a = Java a a". Also, it seems that there are 2 different Java type-constructors in use: "Java a" and "Java a b" (from what I understood both are type constructors as they are used in function signatures and instance definitions). Could you please let me know how I could find the appropriate definition for this?

**Comment by rahulmutt** [comment]
@alanasp The `Java` monad is pretty low-level and defined here: https://github.com/typelead/eta/blob/master/libraries/ghc-prim/GHC/Types.hs#L245 Let me know if you need anymore help.

**Comment by alanasp** [comment]
@rahulmutt Thanks for the prompt answer, it was helpful. I believe the solution to this should be something like this `instance Category Java where
    id = Java $ \ o -> (# o, obj o #);
    f . g = g >- f` Although there is an issue that the signature `(>-) :: (Class b) => Java a b -> Java b c -> Java a c` requires that type `b` is an instance of class `Class`. There is a similar problem for `id` function. Is there a way to fix this without changing the signatures of `Category` functions, which, obviously, is not desirable?

**Comment by rahulmutt** [comment]
You're right. You won't be able to write the instance as it is because the two type parameters of `Java` aren't symmetric. The first parameter has to be a Java class for this to work. In the upcoming revamp of the FFI, we'll try to remove this asymmetry so that this instance can work.

You learn how to insert multiline code snippets [here](https://help.github.com/articles/creating-and-highlighting-code-blocks/).
